### PR TITLE
Inverse kinematic

### PR DESCRIPTION
**Pull request by AndreaCatania**
This is a short video that show IK in action https://youtu.be/hr2x5zWY_AQ

The code should be polished and the main thing to add are the constraints.
The constraints addition is not so important because the algorithm that I've used is FABRIK, and it doesn't need to be constrained to look "reasonably" and this mean that use it will be super easy.

In order to use it you don't need to attach a physical bone to the skeleton. You will need to add it only to apply a constraint (That it's not yer implemented).

## How to use it: 
To use it the first thing to do is create the IK task using this line of code:
```
var goal_transform = Transform()
var right_arm_chain = $Skeleton.create_ik_task("right_arm_bone", "right_hand_bone", goal_transform )
```

Then in the process function you have to call the function to solve the chain:
```
func _process(delta):
	$Skeleton.solve_ik_task( right_arm_chain );
```

Only if the position of task changes update it using:
```
var goal_transform = Transform()
$Skeleton.set_ik_task_goal(right_arm_chain , goal_transform )
```

### CODE REVIEW

**Comment by AndreaCatania** [comment]
The main idea is that provide an API that allow us to use IK on skeleton, then implement some function (or node I hope) allow the animation player to use this API to dosome more "complex" operations (like animation blending with IK, or IK solving on event emitted, etc..).

What you want to do is not so difficult right now but it's very uncomfortable because you have to take care of all things like event managing manually.

In the video that I've posted you can see that the IK solver is running only when the body is inside the button area, when it go outside the solver stop running. Here I could integrate it with an animation but as I said above I should take care of all event  managing manually.

**Comment by mysticfall** [comment]
In Unity, there's `SetIKPositionWeight` and `SetIKRotationWeight` pair with which we can specify how much the IK would affect the animation.

If there was such a function in Godot, I thought it could be a lot easier to deal with such problems that I described above.

For instance, I might add an additional function track in the punching animation which can gradually change the IK weight/influence at the correct moment (i.e. just before the fist hits the target) to achieve the ideal result.

I was wondering if something like this could be implemented with the current binary on/off flag that is included in this PR. And if not, I think it'd be nice to have an API to change IK weight before we can implement more advanced features like the one you described.

Of course, I'm not saying that you should make the change right away, but I just wanted to know if I should open a feature request for it.

The video looks cool, by the way. Can't wait to try it myself :)

**Comment by AndreaCatania** [comment]
The weight essentially is a parameter for blending animation. I'm imagine a scenario where you have an animation tree like this: 

untitled

For this reason I think that we don't have to manage it inside the IK sover, in this way we can take the two things separated and leave to the proper parts makes their job.

I think that manage IK inside the animation tree rather using the code is the best thing to do

**Comment by mysticfall** [comment]
That makes sense. Then I'll wait until we overhaul `AnimationTreePlayer` before 3.1, hopefully with the IK integration feature you suggested. Thanks for the explanation!

**Comment by AndreaCatania** [comment]
@akien-mga Sadly :|  `error: delegating constructors are permitted only in C++11`

Compilation error fixed
Code polished
Improved functions naming
Improved ik API workflow

(Updated "How it works" on first comment)

**Comment by slapin** [comment]
About why I do need to keep some bones rigid in IK.

I have skeleton which I use for animation, it contains limbs of 4 bones, so 2 of bones act as one in case of IK (but that is not a case for animation - some poses simply require this to look ok, i.e. twisiting or muscule bulge simulation, small-scale bending). If IK will require all bones to be IKable, I will have to set up fake bones for IK and manually convert rotations to actual bones in frame loop which will eat-up CPU cycles. Swichig meshes/skeletons are also a problem (i.e. I want to use IK to fixup animation, not just posing, so I will need full skeleton for that).

**Comment by AndreaCatania** [comment]
Talk about full skeleton IK is wrong. If you try this pr you will see that you have to create an IK task where you have to set root and tip bones. If the two bones to handle the muscle are outside of this chain its basis will not be touched.

When you create the task the chain is composed only of direct bones, all other appended bones doesn't partecipate to the solving and its basis is not touched 

 


If this is not your case please send me an image that represent your skeleton

**Comment by slapin** [comment]
Well, the chain is constructed of 4 bones, [base]upperleg01.L,
upperleg02.L, [kneel is here] lowerleg01.L, lowerleg02.L[tip]
So in this case we need to implement IK over 2 bones - upperleg01 and
lowerleg01 but we neeed to consider whole chain length.

**Comment by LeonardMeagher2** [comment]
@slapin are you saying the "kneel" is the parent to lowerleg01.L?
I'd imagine upperleg02.L should be a parent to both lowerleg01.L and "kneel"

**Comment by slapin** [comment]
"Kneel" is not a bone, just the reference to understand the layout

Bones are: upperleg01.L upperleg02.L lowerleg01.L lowerleg02.L

upperleg01 is start of chain
lowerleg02 is end of chain

kneel is between upperleg02 and lowerleg01 which splits the limb into

upper part - upperleg01.L upperleg02.L
lower part - lowerleg01.L lowerleg02.L

each bening as single bone by IK and independently by animation.

**Comment by pwab** [comment]
As discussed with @TwistedTwigleg in godotengine/godot-demo-projects/issues/211 I also tried to impement FABRIK but for 2d IK chains. Is your implementation somehow transferable for 2d skeletons, @AndreaCatania?

**Comment by AndreaCatania** [comment]
Yes it's possible, it's not too difficult take the code that I written and change Vector3 with Vector2 etc.. Feel free to do it if you want

**Comment by Zireael07** [comment]
Is this merge-ready or is it on hold like the soft body PR?

**Comment by AndreaCatania** [comment]
It works, but reduz said that I need to add a function to blend with current animation pose. I think that I can do it today. I let you know

**Comment by AndreaCatania** [comment]
I can't do it today I will finish it this weekend sorry

**Comment by AndreaCatania** [comment]
I've implemented the blending functionality and this PR is ready to be merged.

Now is possible to use it with animation player so you can use IK easily with your animations.
Here a video: https://youtu.be/PXhF78c6wRs
and here the project to test it, and see how I've implemented it using animation player: 
[ik.zip](https://github.com/godotengine/godot/files/2086875/ik.zip)

**Comment by AndreaCatania** [comment]
I've implemented the new node SkeletonIK as @reduz proposed.

Now is possible to set a **root** and **tip** bone and a target transform or a spatial node as target directly in the SkeletonIK node.
Then by calling the method **start** now the IK will be solved automatically and if you set a node as target the IK will follow it.

And also is possible to set the interpolation value, that can be changed at runtime, check this project to test it: [ik.zip](https://github.com/godotengine/godot/files/2108912/ik.zip)

Let me know what you think about!

**Comment by AndreaCatania** [comment]
The branch is `AndreaCatania:ik`

> Also did not you consider trivial cases like 2 bones with target node and pole?

What do you mean?


> Also as was discussed before, was there any ideas regarding gaps in bone chain (for several bones acting as single bone)?

This is really particular case, and could be addressed later when this PR will be merged. A solution could be trivially use two separate SkeletonIK (chain) but should be tested in order to understand the use case and if the result is correct

**Comment by slapin** [comment]
I mean 2 bone IK is special case and optimization. You can directly solve 2
bone IK and not run any algorithms like CCD/FABRIK, etc.
No iterations, just the result.
This allows to have complicated IK setups cheaply (and be simple for simple
set ups). I hope this is not too hard to understand.
The generic IK case it too slow for real practical (runtime) use,
especially for complex scenes. Most of practical set ups for aim (arms) and
uneven ground (feet) use 2 bone IK.

**Comment by LeonardMeagher2** [comment]
If you look up the law of cosines, it can solve 2 bone ik pretty easily

**Comment by AndreaCatania** [comment]
Oh, you was talking about optimization... Now I see what you are talking about... Anyway I will see if it worth implement it because FABRIK is not so heavy algorithm expecially in case you have just two bones.

Do you have a prove of its bad performance insue that I can check?

**Comment by reduz** [comment]
I'm on Andrea on this one, I don't think this is a case worth optimizing.

**Comment by slapin** [comment]
Well, what is bad and good is basically a comparison. My target is 200
characters per scene.
I did not try your implementation, but with some other implementations I
find that I can have
5 characters per scene with CCD and 20 with FABRIK (minus heavy animation
stuff which sucks).
With 2-bone IK I can have 200 characters on the same hardware (running
animation).
Of cource LODs and animation lods and manual LODs are utilized at the same
extent on all the scenes.

**Comment by slapin** [comment]
with FABRIK you will need a few iterations in best case, while with 2-bone
IK you will
give direct single answer without iterations (and that is 5 lines of code).

**Comment by AndreaCatania** [comment]
In case of 200 characters I think that for 90% of these you don't need IK at all, anyway I will add some functions in order to expose max iterations and bias so you can control how precision should be.

Also is interesting for me test this scene with 200 character to see what we are talking about

**Comment by slapin** [comment]
For me the most common scenario where I need IK + 200 people is
like
https://www.pixoto.com/images-photography/people/street-and-candids/crowded-spanish-steps-in-rome-5019323078279168
where errors
in feet motion are visible and at the same time you don't want to make it
fixed - that will be easily noticed as repeatitive, so procedural
animation is required.
As Ubisoft can do this I think Godot should too...
https://www.youtube.com/watch?v=kOv5ixy7Y8w

Anyway ease of LOD usage with IK and animations should be required to do
true massive crowds like in Ubisoft games.
(they also replace distant characters with much simpler meshes with much
simpler skeletons, but that is another story).

**Comment by AndreaCatania** [comment]
In this picture the error on the feet are visible only on the first line of people sitdown on the wall, and for all others you don't need the IK to fix it.

The same rule goes with Ubisoft video, do you think that all people that are so faraway that seems just a texture have the IK active? I don't think so... you need a perfect IK only for nearest person and should be activated only when required so when a person is out of focus but still near and visible you need to solve IK with just a good precision, but for all other you don't need to solve..

Also when you resolve the IK one time you don't need to re solve it again if you don't move the bones.

Anyway I think that for talk about what we have to change to optimize we first need a scene that runs at low FPS to check it... because in this moment we are talk about a real picture and what you could need to make it in godot

**Comment by AndreaCatania** [comment]
The IK had the problem that when you wanted to make it works with the animation player, the animation player override the bone poistion if executed after IK solver.

In order to make it works correctly I've moved the execution code from process_internal to process, so I'm always sure that the IK solver is processed always after all animation players.

This doesn't make any differences on Editor side and is always possible to override the _process function by a script.

**Comment by hpvb** [comment]
@andreacatania are you happy with the current state of this PR? I think it should be merged if you are happy.

**Comment by AndreaCatania** [comment]
No is required to make a little change that can't be done without merging
this pr : #19888

**Comment by karroffel** [comment]
I mentioned this before in a private discussion, but I'll leave it here so others can see it too.

I tested this PR, and while it works really well, the interface and way to use it is quite different than in pretty much all other software that let's the user use IK.

With this PR, a target transform is applied to the bone on the tip of the chain, then the rest of the chain "follows". That means that the *rotation* of the target has to match the one of the desired pose.

This is a lot harder to get right.

The usual approach is to have a target (that only contributes position by default, rotation opt-in) and a *pole* that dictates how the chain will bend.

From what I have seen, there is an official IK demo that adds a "middle bone helper" (so a pole basically) on top of the FABRIK algorithm.

https://github.com/godotengine/godot-demo-projects/blob/c6b481aaf0cfb677b182b0e3711702c15b0cb413/3d/ik/addons/sade/ik_fabrik.gd

Maybe this PR can be modified so it supports the more "standard" target+pole workflow.

File: scene/animation/skeleton_ik.h [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+	// The goal of chain should be always in local space
+	static void set_goal(Task *p_task, const Transform &p_goal);
+	static void make_goal(Task *p_task, const Transform &p_inverse_transf, real_t blending_delta);
+	static void solve(Task *p_task, real_t blending_delta);
+};
+
+class SkeletonIK : public Node {
+	GDCLASS(SkeletonIK, Node);
+
+	String root_bone;
```
Comment (reduz) [inline scene/animation/skeleton_ik.h]: by the way, I think this should be StringName, not String

File: scene/animation/skeleton_ik.h [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+	void set_tip_bone(const String &p_tip_bone);
+	const String &get_tip_bone() const;
+
+	void set_interpolation(real_t p_interpolation);
+	real_t get_interpolation() const;
+
+	void set_target(const Transform &p_target);
+	const Transform &get_target() const;
+
+	void set_target_node_override(const NodePath &p_node);
```
Comment (reduz) [inline scene/animation/skeleton_ik.h]: I dont think this should be called override, if set it should be obvious (And docummented) that takes priority over a transform

File: scene/animation/skeleton_ik.h [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+	void set_root_bone(const String &p_root_bone);
+	const String &get_root_bone() const;
+
+	void set_tip_bone(const String &p_tip_bone);
+	const String &get_tip_bone() const;
+
+	void set_interpolation(real_t p_interpolation);
+	real_t get_interpolation() const;
+
+	void set_target(const Transform &p_target);
```
Comment (reduz) [inline scene/animation/skeleton_ik.h]: I would call this set_target_transform

File: scene/animation/skeleton_ik.h [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+	const Transform &get_target() const;
+
+	void set_target_node_override(const NodePath &p_node);
+	NodePath get_target_node_override();
+
+	void set_min_distance(real_t p_min_distance);
+	real_t get_min_distance() const { return min_distance; }
```
Comment (reduz) [inline scene/animation/skeleton_ik.h]: I dont quite understand what this is for

Comment (AndreaCatania) [reply scene/animation/skeleton_ik.h]: This is used to control the precision of IK, it tells the minimum distance between target and tip before consider the process done.

File: scene/animation/skeleton_ik.h (hunk 1 above)
Comment (AndreaCatania) [reply scene/animation/skeleton_ik.h]: I changed it, but seems to me that it is not better, because the skeleton internally store bone names as String

Comment (AndreaCatania) [reply scene/animation/skeleton_ik.h]: I've tried to set StringName anyway, but I got compilation error..

File: scene/animation/skeleton_ik.h (hunk 2 above)
Comment (AndreaCatania) [reply scene/animation/skeleton_ik.h]: Done

File: scene/animation/skeleton_ik.h (hunk 3 above)
Comment (AndreaCatania) [reply scene/animation/skeleton_ik.h]: done

**Comment by AndreaCatania** [comment]
@reduz  I've a problem with process_internal priority.

The execution order seems random, (or better seems related to the area of last mouse click), when I start the animation and IK in the editor the IK sometimes is processed before the animation player and sometimes later...
Even if the priority of AnimatorPlayer is 0 and should be executed before SkeletonIK that has the priority 1.

check video: https://www.youtube.com/watch?v=yBliluCY6M0 

Ther's something other that I need to address to properly implement the priority?

**Comment by akien-mga** [comment]
Needs a rebase to fix conflicts in skeleton.cpp.

**Comment by AndreaCatania** [comment]
Rebased

File: scene/main/node.h [hunk 5]
```diff
@@ -72,7 +72,7 @@ class Node : public Object {
 
 	struct ComparatorWithPriority {
 
-		bool operator()(const Node *p_a, const Node *p_b) const { return p_b->has_priority_higher_than(p_a) || p_b->is_greater_than(p_a); }
```
Comment (reduz) [inline scene/main/node.h]: Just request the priority (integer) and compare integers here

Comment (reduz) [reply scene/main/node.h]: no need to add a function that no one else will use

**Review by reduz — CHANGES_REQUESTED** [review]

**Pull request merged**